// siesta-lite/src/nodejs/launcher/NodeJsPageRpc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
//---------------------------------------------------------------------------------
var NodeJsIpcRpc;
(function (NodeJsIpcRpc) {
    NodeJsIpcRpc[NodeJsIpcRpc["FunctionCall"] = 0] = "FunctionCall";
})(NodeJsIpcRpc = exports.NodeJsIpcRpc || (exports.NodeJsIpcRpc = {}));
//---------------------------------------------------------------------------------
var NodeJsPageRpcFunctionCall = /** @class */ (function () {
    function NodeJsPageRpcFunctionCall(config) {
        this.type = NodeJsIpcRpc.FunctionCall;
        Object.assign(this, config);
    }
    NodeJsPageRpcFunctionCall.prototype.toJson = function () {
        var json = {
            type: this.type,
            func: this.func.toString()
        };
        if (this.args)
            json.args = this.args;
        return json;
    };
    return NodeJsPageRpcFunctionCall;
}());
exports.NodeJsPageRpcFunctionCall = NodeJsPageRpcFunctionCall;
//---------------------------------------------------------------------------------
var NodeJsIpcClienState;
(function (NodeJsIpcClienState) {
    NodeJsIpcClienState["Ready"] = "CLIENT_READY";
})(NodeJsIpcClienState = exports.NodeJsIpcClienState || (exports.NodeJsIpcClienState = {}));
