// siesta-lite/webpack.config.js
const
    webpack                 = require('webpack'),
    path                    = require('path')


module.exports = (env) => {
    if (!env) env = {}

    const simulator = {
        mode    : 'none',
        entry   : './src/entries/simulator_client.js',
        output  : {
            filename        : 'simulator-client-all.js',
            path            : path.resolve(__dirname, 'build/bundles'),

            library         : 'simulator_client',
            libraryTarget   : 'assign',
            umdNamedDefine  : true
        }
    }


    const config = []

    config.push(
        simulator
        // ,
        // {
        //     entry   : './src/entries/util.js',
        //     output  : {
        //         filename        : 'util-all.js',
        //         path            : path.resolve(__dirname, 'build/bundles'),
        //
        //         library         : 'util',
        //         libraryTarget   : 'assign',
        //         umdNamedDefine  : true
        //     }
        // }
    )

    return config
}
